Add the runtime flag and basic implementation of :dir

rniwa opened the pull request:
#### 143e43df80c213c4d3e017c55f9de94d368de4c4
<pre>
Add the runtime flag and basic implementation of :dir
<a href="https://bugs.webkit.org/show_bug.cgi?id=243090">https://bugs.webkit.org/show_bug.cgi?id=243090</a>

Reviewed by Cameron McCormack.

Added the rudimentary support for :dir under a new runtime flag.

* LayoutTests/TestExpectations: Removed failing expectations from now passing tests.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-selector-auto-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-selector-querySelector-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-html-input-dynamic-text-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir01-expected.txt: Ditto.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml: Added a runtime flag.
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesDirPseudoClass): Added.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::computeDirectionality const): Added.
* Source/WebCore/html/HTMLElement.h:

Canonical link: <a href="https://commits.webkit.org/252737@main">https://commits.webkit.org/252737@main</a>
</pre>

heycam commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -88,6 +88,7 @@ class HTMLElement : public StyledElement {
 
+    TextDirection computeDirectionality() const;
     bool hasDirectionAuto() const;
     TextDirection directionalityIfhasDirAutoAttribute(bool& isAuto) const;
```
Pre-existing issue, and unrelated to your patch, but the "has" in this function name should probably be "Has".

heycam commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -248,6 +248,27 @@ ALWAYS_INLINE bool matchesLangPseudoClass(const Element& element, const Vector<A
     return false;
 }
 
+#if ENABLE(CSS_SELECTORS_LEVEL4)
```
Do you know why `CSS_SELECTORS_LEVEL4` is enabled only for Cocoa platforms? Seems strange to me that some core CSS functionality is not enabled everywhere.

rniwa replied:
No idea. It's probably some historical artifact. We should probably enable it everywhere and remove the compile flag altogether.

rniwa replied:
Tracking that in https://bugs.webkit.org/show_bug.cgi?id=243092

heycam reviewed (approved):
Looks like a good start. I guess the main missing thing here is handling dynamic changes.

rniwa committed:
Add the runtime flag and basic implementation of :dir
https://bugs.webkit.org/show_bug.cgi?id=243090

Reviewed by Cameron McCormack.

Added the rudimentary support for :dir under a new runtime flag.

* LayoutTests/TestExpectations: Removed failing expectations from now passing tests.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-selector-auto-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-selector-querySelector-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-html-input-dynamic-text-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir01-expected.txt: Ditto.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml: Added a runtime flag.
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesDirPseudoClass): Added.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::computeDirectionality const): Added.
* Source/WebCore/html/HTMLElement.h:

Canonical link: https://commits.webkit.org/252737@main
